Lib/test/test_threading.py

```diff
@@ -256,32 +254,60 @@ def test_various_ops_large_stack(self):
 
     def test_foreign_thread(self):
         # Check that a "foreign" thread can use the threading module.
+        dummy_thread = None
+        error_message = None
         def f(mutex):
             # Calling current_thread() forces an entry for the foreign
             # thread to get made in the threading._active map.
-            threading.current_thread()
-            mutex.release()
+            try:
+                nonlocal dummy_thread
+                nonlocal error_message
+                dummy_thread = threading.current_thread()
+                tid = dummy_thread.ident
+                def verify_conditions():
+                    if tid not in threading._active:
+                        return f'Expected {tid} to be in {threading._active}'
+                    if not isinstance(dummy_thread, threading._DummyThread):
+                        return f'Expected {dummy_thread} to be an instance of threading._DummyThread.'
+                    found = threading._active.get(tid)
+                    if found is not dummy_thread:
+                        return f'Expected {found} to be {dummy_thread}.'
+                    if not dummy_thread.is_alive():
+                        # gh-29376
+                        return 'Expected _DummyThread to be considered alive.'
+                    r = repr(threading._active[tid])
+                    if '_DummyThread' not in r:
+                        return f'Expected "_DummyThread" to be in {r}.'
+                    return None
+                error_message = verify_conditions()
```

Comment:
If the code above raises an exception, `error_message` will keep to be None.

Maybe catch an exception and set it to a nonlocal variable? You can also use `self.assert*()` for checks.
```py
try:
    # verify
    self.assertIn(tid, threading._active)
    ...
except BaseException as e:
    nonlocal error
    error = e
finally:
    mutex.release()

...
if error is not None:
    raise error
```

Reply:
Done